api/bag_dataset.h

```diff
@@ -0,0 +1,192 @@
+#ifndef BAG_DATASET_H
+#define BAG_DATASET_H
+
+#include "bag_compounddatatype.h"
+#include "bag_compoundlayerdescriptor.h"
+#include "bag_config.h"
+#include "bag_descriptor.h"
+#include "bag_exceptions.h"
+#include "bag_fordec.h"
+#include "bag_layer.h"
+#include "bag_metadata.h"
+#include "bag_trackinglist.h"
+#include "bag_types.h"
+#include "bag_vrtrackinglist.h"
+
+#include <functional>
+#include <memory>
+#include <numeric>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+
+namespace H5 {
+
+class H5File;
+
+}   //namespace H5
+
+namespace BAG {
+
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(disable: 4251)  // std classes do not have DLL-interface when exporting
+#pragma warning(disable: 4275)  // non-DLL-interface class used as base class
+#endif
+
+//! The interface for the BAG.
+/*!
+    This is the BAG Dataset.  It is responsible for creating or reading a BAG.
+    It contains the layers and metadata.  It has a descriptor for general
+    details.
+
+    It is the only way layers can be created.  It is meant to be the interface
+    to the BAG, and answer questions about said BAG.
+*/
+class BAG_API Dataset final
+    : public std::enable_shared_from_this<Dataset>
+
+{
+public:
+    static std::shared_ptr<Dataset> open(const std::string &fileName,
+        OpenMode openMode);
+
+    static std::shared_ptr<Dataset> create(const std::string &fileName,
+        Metadata&& metadata, uint64_t chunkSize = 100,
+        int compressionLevel = 5);
+
+    void close();
+
+    Dataset(const Dataset&) = delete;
+    Dataset(Dataset&&) = delete;
+
+    Dataset& operator=(const Dataset&) = delete;
+    Dataset& operator=(Dataset&&) = delete;
+
+    bool operator==(const Dataset &rhs) const noexcept {
+        return m_pH5file == rhs.m_pH5file &&
+               m_layers == rhs.m_layers &&
+               m_pMetadata == rhs.m_pMetadata &&
+               m_pTrackingList == rhs.m_pTrackingList &&
+               m_descriptor == rhs.m_descriptor &&
+               m_pVRTrackingList == rhs.m_pVRTrackingList;
+    }
+
+    bool operator!=(const Dataset &rhs) const noexcept {
+        return !(rhs == *this);
+    }
+
+    Layer& getLayer(uint32_t id) &;
+    const Layer& getLayer(uint32_t id) const &;
+    Layer* getLayer(LayerType type, const std::string& name) &;
```

Comment:
Since name is optional per documentation comments in the implementation, add an empty string as a default value.

Reply:
@rolker Unless I am mistaken, `name` is required for these public `getLayer` functions. It is option for the internal helper function `getLayer`. Closing for now.